File: src/main/clojure/czlab/basal/ebus.clj
;; Copyright (c) 2013-2017, Kenneth Leung. All rights reserved.
;; The use and distribution terms for this software are covered by the
;; Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
;; which can be found in the file epl-v10.html at the root of this distribution.
;; By using this software in any fashion, you are agreeing to be bound by
;; the terms of this license.
;; You must not remove this notice, or any other, from this software.

(ns ^{:doc ""
      :author "Kenneth Leung"}

  czlab.basal.ebus

  (:require [clojure.string :as cs]))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;(set! *warn-on-reflection* true)

(def ^:private _SEED (atom 1))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn- splitTopic "" [topic] (if (and (string? topic)
                                      (not-empty topic))
                               (->> (cs/split topic #"/")
                                    (filterv #(if (> (count %) 0) %)))
                               []))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn- nextSEQ "" [] (let [n @_SEED] (swap! _SEED inc) n))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defprotocol EventBus
  ""
  ())

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn- mkSubSCR
  "" [topic repeat? listener args]
  {:pre [(fn? listener)]}
  {:id (keyword (str "s#" (nextSEQ)))
   :repeat? repeat?
   :action listener
   :topic topic
   :args args
   :status (long-array 1 1)})

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; nodes - children
;; subscribers
(defn- mkTreeNode "" [] {:nodes {} :subcs {}})

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn- addOneSub "" [node sub]
  (update-in node [:subcs] assoc (:id sub) sub))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn- remOneSub "" [node sub]
  (update-in node [:subcs] dissoc (:id sub)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn- addOneTopic "" [top {:keys [topic] :as sub}]
  (let [path (splitTopic topic)]
    (-> (update-in top path addOneSub sub)
        (update-in [:subcs] assoc (:id sub) sub))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn- addTopic "" [root sub] (swap! root addOneTopic sub))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; for each topic, subscribe to it.
(defn- listen "" [root repeat? topics listener more]
  (->> (-> (or topics "") cs/trim (cs/split #"\s+"))
       (filter #(if (> (count %) 0) %))
       (mapv #(addTopic root
                        (mkSubSCR % repeat? listener more)))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn- delOneTopic "" [top {:keys [topic] :as sub}]
  (let [path (splitTopic topic)]
    (-> (update-in top path remOneSub sub)
        (update-in [:subcs] dissoc (:id sub)))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn- unSub "" [root sub] (swap! root delOneTopic sub))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn- doPub "" [branch pathTokens topic msg]
  (let [{:keys [nodes subcs]} branch
        [p & more] pathTokens
        cur (get nodes p)
        s1 (get nodes "*")
        s1c (:nodes s1)
        s2 (get nodes "**")]
    (if s2
      (run s2 topic msg))
    (if s1
      (cond
        (and (empty? more)
             (empty? s1c))
        (run s1 topic msg)
        (or (and (empty? s1c)
                 (not-empty more))
            (and (empty? more)
                 (not-empty s1c)))
        nil
        :else
        (doPub s1 more topic msg)))
    (if cur
      (if (not-empty more)
        (doPub cur more topic msg)
        (run cur topic msg)))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn- run "" [{:keys [subcs]} topic msg]
  (doseq [z subcs
          :let [{:keys [repeat? action status extraArgs]} z]
          :when (pos? (aget ^longs status 0))]
    (apply action (concat [topic msg] extraArgs))
    ;;if one time only, turn off flag
    (if-not repeat?
      (aset ^longs status 0 -1))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn eventBus<> "" []
  (let [state (atom (mkTreeNode))]
    (reify EventBus
      ;; Subscribe to 1+ topics, returning a list of subscriber handles.
      ;; topics => "/hello/*  /goodbye/*"
      ;; @param {String} topics - space separated if more than one.
      ;; @return {Array.String} - subscription ids
      (once [_ topics listener & args]
        (or (listen state false topics listener args) []))
      ;; subscribe to 1+ topics, returning a list of subscriber handles.
      ;; topics => "/hello/*  /goodbye/*"
      ;; @param {String} topics - space separated if more than one.
      ;; @param {Function} selector
      ;; @return {Array.String} - subscription ids.
      (on [_ topics listener & args]
        (or (listen state true topics listener args) []))
      ;; Trigger event on this topic.
      ;; @param {String} topic
      ;; @param {Object} msg
      ;; @return {Boolean}
      (fire [_ topic msg]
        (let [tokens (splitTopic topic)]
          (if (not-empty tokens)
            (doPub @state tokens topic msg))))
      ;; Resume actions on this handle.
      ;; @memberof module:cherimoia/ebus~RvBus
      ;; @method resume
      ;; @param {Object} - handler id
      (resume [_ handle]
        (let [sub (get (:subcs @state) handle)
              r? (true? (:repeat? sub))
              st (if sub (:status sub))
              sv (if st (aget ^longs st 0) -911)]
          (if (= 0 sv)
            (aset ^longs st 0 1))))
      ;; Pause actions on this handle.
      ;; @memberof module:cherimoia/ebus~RvBus
      ;; @method pause
      ;; @param {Object} - handler id
      (pause [_ handle]
        (let [sub (get (:subcs @state) handle)
              st (if sub (:status sub))
              sv (if st (aget ^longs st 0) -911)]
          (if (pos? sv)
            (aset ^longs st 0 0))))
      ;; Stop actions on this handle.
      ;; @memberof module:cherimoia/ebus~RvBus
      ;; @method off
      ;; @param {Object} - handler id
      (off [_ handle]
        (if-some [sub (get (:subcs @state) handle)]
          (unSub state sub)))
      ;; Remove all subscribers.
      (removeAll [_] (reset! state (mkTreeNode))))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;EOF

File: src/main/cljs/czlab/tzone/ebus.js
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Copyright (c) 2013-2016, Kenneth Leung. All rights reserved.
*/


"use strict";/**
 * @requires cherimoia/skaro
 * @module cherimoia/ebus
 */
import sjs from "cherimoia/skaro";

let R = sjs.ramda,
undef,
_SEED=0;

//////////////////////////////////////////////////////////////////////////////
const mkSubSCR = (topic, selector, context, repeat, args) => {
  return {
    target:  !!context ? context : null,
    id: "sub#" + Number(++_SEED),
    repeat: sjs.boolify(repeat),
    args: args || [],
    action: selector,
    topic: topic,
    active: true
  };
}

//////////////////////////////////////////////////////////////////////////////
const mkTreeNode = (root) => {
  return root ? { tree: {} } : {
    tree: {},  // children - branches
    subs: []   // subscribers
  };
}

//////////////////////////////////////////////////////////////////////////////
const _listen = function(repeat, topics, selector, target, more) {
  const ts= topics.trim().split(/\s+/),
  // for each topic, subscribe to it.
  rc= R.map(t => {
    return _addSub.call(this, repeat,t,selector,target,more);
  }, ts);
  return R.reject( z => { return z.length===0; }, rc);
}

//////////////////////////////////////////////////////////////////////////////
const _doSub = (node,token) => {
  if (! sjs.hasKey(node.tree, token)) {
    node.tree[token] = mkTreeNode();
  }
  return node.tree[token];
}

//////////////////////////////////////////////////////////////////////////////
const _addSub = function(repeat, topic, selector, target, more) {
  let tkns= this.splitTopic(topic),
  rcid='';
  if (tkns.length > 0) {
    const rc= mkSubSCR(topic, selector, target, repeat, more),
    node= R.reduce((memo, z) => {
      return _doSub(memo,z);
    },
    this.root, tkns);

    this.subs[rc.id] = rc;
    node.subs.push(rc);
    rcid= rc.id;
  }
  return rcid;
}

//////////////////////////////////////////////////////////////////////////////
const _unSub = function(node, tokens, pos, sub) {
  if (! sjs.echt(node)) { return; }
  if (pos < tokens.length) {
    const k= tokens[pos],
    cn= node.tree[k];
    _unSub.call(this, cn, tokens, pos+1, sub);
    if (R.keys(cn.tree).length === 0 &&
        cn.subs.length === 0) {
      delete node.tree[k];
    }
  } else {
    pos = -1;
    R.find( z => {
      pos += 1;
      if (z.id === sub.id) {
        delete this.subs[z.id];
        node.subs.splice(pos,1);
        return true;
      }
    }, node.subs);
  }
}

//////////////////////////////////////////////////////////////////////////////
const _doPub = function(topic, node, tokens, pos, msg) {
  if (! sjs.echt(node)) { return false; }
  let rc=false;
  if (pos < tokens.length) {
    rc = rc || _doPub.call(this,topic, node.tree[ tokens[pos] ], tokens, pos+1, msg);
    rc = rc || _doPub.call(this,topic, node.tree['*'], tokens, pos+1,msg);
  } else {
    rc = rc || _run.call(this,topic, node,msg);
  }
  return rc;
}

//////////////////////////////////////////////////////////////////////////////
const _run = function(topic, node, msg) {
  const cs= !!node ? node.subs : [];
  let purge=false,
  rc=false;

  R.forEach( z => {
    if (z.active &&
        sjs.echt(z.action)) {
      // pass along any extra parameters, if any.
      z.action.apply(z.target, [msg,topic].concat(z.args));
      // if once only, kill it.
      if (!z.repeat) {
        delete this.subs[z.id];
        z.active= false;
        z.action= null;
        purge=true;
      }
      rc = true;
    }
  }, cs);

  // get rid of unwanted ones, and reassign new set to the node.
  if (purge && cs.length > 0) {
    node.subs= R.filter( z => {
      return z.action ? true : false;
    }, cs);
  }

  return rc;
}

//////////////////////////////////////////////////////////////////////////////
/** @class RvBus */
class RvBus extends sjs.ES6Claxx {
  /**
   * Subscribe to 1+ topics, returning a list of subscriber handles.
   * topics => "/hello/*  /goodbye/*"
   * @memberof module:cherimoia/ebus~RvBus
   * @method once
   * @param {String} topics - space separated if more than one.
   * @param {Function} selector
   * @param {Object} target
   * @return {Array.String} - subscription ids
   */
  once(topics, selector, target /*, more args */) {
    const rc= _listen.call(this,false, topics,
                            selector, target,
                            sjs.dropArgs(arguments,3));
    return sjs.echt(rc) ? rc : [];
  }

  /**
   * subscribe to 1+ topics, returning a list of subscriber handles.
   * topics => "/hello/*  /goodbye/*"
   * @memberof module:cherimoia/ebus~RvBus
   * @method on
   * @param {String} topics - space separated if more than one.
   * @param {Function} selector
   * @param {Object} target
   * @return {Array.String} - subscription ids.
   */
  on(topics, selector, target /*, more args */) {
    const rc= _listen.call(this,true, topics,
                            selector, target,
                            sjs.dropArgs(arguments,3));
    return sjs.echt(rc) ? rc : [];
  }

  /**
   * Trigger event on this topic.
   * @memberof module:cherimoia/ebus~RvBus
   * @method fire
   * @param {String} topic
   * @param {Object} msg
   * @return {Boolean}
   */
  fire(topic, msg) {
    let tokens= this.splitTopic(topic),
    rc=false;
    if (tokens.length > 0 ) {
      rc= _doPub.call(this,topic,
                       this.root, tokens, 0, msg || {} );
    }
    return rc;
  }

  /**
   * @method splitTopic
   * @protected
   */
  splitTopic(topic) {
    return sjs.safeSplit(topic,'/');
  }

  /**
   * Resume actions on this handle.
   * @memberof module:cherimoia/ebus~RvBus
   * @method resume
   * @param {Object} - handler id
   */
  resume(handle) {
    const sub= this.subs[handle];
    if (!!sub) {
      sub.active=true;
    }
  }

  /**
   * Pause actions on this handle.
   * @memberof module:cherimoia/ebus~RvBus
   * @method pause
   * @param {Object} - handler id
   */
  pause(handle) {
    const sub= this.subs[handle];
    if (!!sub) {
      sub.active=false;
    }
  }

  /**
   * Stop actions on this handle.
   * Unsubscribe.
   * @memberof module:cherimoia/ebus~RvBus
   * @method off
   * @param {Object} - handler id
   */
  off(handle) {
    const sub= this.subs[handle];
    if (!!sub) {
      _unSub.call(this,this.root,
                   this.splitTopic(sub.topic), 0, sub);
    }
  }

  /**
   * @method iniz
   * @private
   */
  iniz() {
    this.root= mkTreeNode(true);
    this.subs = {};
  }

  /**
   * Remove all subscribers.
   * @memberof module:cherimoia/ebus~RvBus
   * @method removeAll
   */
  removeAll() {
    this.iniz();
  }

  /**
   * @method constructor
   * @private
   */
  constructor() {
    super();
    this.iniz();
  }

};
//////////////////////////////////////////////////////////////////////////////
/**
 * @class EventBus
 */
class EventBus extends RvBus {
  /**
   * Trigger event on this topic.
   * @memberof module:cherimoia/ebus~EventBus
   * @method fire
   * @param {String} topic
   * @param {Object} msg
   * @return {Boolean}
   */
  fire(topic, msg) {
    return _run.call(this, topic,
                     this.root.tree[topic], msg || {});
  }

  /**
   * @method splitTopic
   * @protected
   */
  splitTopic(topic) { return [topic]; }

  /**
   * @method constructor
   * @private
   */
  constructor() {
    super();
  }
}

/** @alias module:cherimoia/ebus */
const xbox= /** @lends xbox# */{
  /**
   * @method reifyRvBus
   * @return {RvBus}
   */
  reifyRvBus() {
    return new RvBus();
  },
  /**
   * @method reify
   * @return {EventBus}
   */
  reify() {
    return new EventBus();
  },
  /**
   * @property {EventBus} EventBus
   */
  EventBus: EventBus,
  /**
   * @property {RvBus} RvBus
   */
  RvBus: RvBus
};

sjs.merge(exports, xbox);
/*@@
return xbox;
@@*/

//////////////////////////////////////////////////////////////////////////////
//EOF

